spacy/tests/test_visualization.py

```diff
@@ -0,0 +1,1080 @@
+            "dep_",
+            value_dependent_bg_colors={"dep": 11},
+        ).render(doc[2])
+        == "\x1b[48;5;11mdep\x1b[0m"
+        if supports_ansi
+        else "dep"
+    )
+
+
+def test_visualization_render_native_attribute_missing(en_vocab):
+    doc = Doc(
+        en_vocab,
+        words=[
+            "I",
+            "saw",
+            "a",
+            "horse",
+            "yesterday",
+            "that",
+            "was",
+            "injured",
+            ".",
+        ],
+        heads=[1, None, 3, 1, 1, 7, 7, 3, 1],
+        deps=["dep"] * 9,
+    )
+    with pytest.raises(AttributeError):
+        AttributeFormat("depp").render(doc[2])
+
+
+def test_visualization_render_custom_attribute_str(en_vocab):
+    doc = Doc(
+        en_vocab,
+        words=[
+            "I",
+            "saw",
+            "a",
+            "horse",
+            "yesterday",
+            "that",
+            "was",
+            "injured",
+            ".",
+        ],
+        heads=[1, None, 3, 1, 1, 7, 7, 3, 1],
+        deps=["dep"] * 9,
+    )
+    Token.set_extension("test", default="tested", force=True)
+    assert AttributeFormat("_.test").render(doc[2]) == "tested"
+
+
+def test_visualization_render_nested_custom_attribute_str(en_vocab):
+    doc = Doc(
+        en_vocab,
+        words=[
+            "I",
+            "saw",
+            "a",
+            "horse",
+            "yesterday",
+            "that",
+            "was",
+            "injured",
+            ".",
+        ],
+        heads=[1, None, 3, 1, 1, 7, 7, 3, 1],
+        deps=["dep"] * 9,
+    )
+
+    class Test:
+        def __init__(self):
+            self.inner_test = "tested"
+
+    Token.set_extension("test", default=Test(), force=True)
+    assert AttributeFormat("_.test.inner_test").render(doc[2]) == "tested"
+
+
+def test_visualization_render_custom_attribute_missing(en_vocab):
+    doc = Doc(
+        en_vocab,
+        words=[
+            "I",
+            "saw",
+            "a",
+            "horse",
+            "yesterday",
+            "that",
+            "was",
+            "injured",
+            ".",
+        ],
+        heads=[1, None, 3, 1, 1, 7, 7, 3, 1],
+        deps=["dep"] * 9,
+    )
+    with pytest.raises(AttributeError):
+        AttributeFormat("._depp").render(doc[2])
+
+
+def test_visualization_render_permitted_values(en_vocab):
+    doc = Doc(
+        en_vocab,
+        words=[
+            "I",
+            "saw",
+            "a",
+            "horse",
+            "yesterday",
+            "that",
+            "was",
+            "injured",
+            ".",
+        ],
+        heads=[1, None, 3, 1, 1, 7, 7, 3, 1],
+        deps=["dep"] * 9,
+    )
+    attribute_format = AttributeFormat("head.i", permitted_values=(3, 7))
+    assert [attribute_format.render(token) for token in doc] == [
+        "",
+        "",
+        "3",
+        "",
+        "",
+        "7",
+        "7",
+        "3",
+        "",
+    ]
```

Comment:
In these cases, when the `black` formatting pushes each item onto one line, it could make sense to define the expected list upfront in a var, then compare to that var. In general I think it would be good to compress some of these tests a bit to make things a bit more readible.